### PR TITLE
Refactor TLS setup code into a separate file and consolidate logic between the StatefulSet and exporter Deployment

### LINKED ISSUE #308: Refactor TLS code
The code that enables TLS for Solr and Prometheus exporter pods has evolved over the past several months, and still has some additional changes we want to add (e.g. https://github.com/apache/solr-operator/issues/300). Let's clean-up the TLS code before adding more features as it's become a bit hard to follow the logic and is infused across too many places. This will also help reduce the size and complexity of the `GenerateStatefulSet` method in `solr_util.go`.

### PR DESCRIPTION
**Pull request by thelabdude**
Fixes #308 

Consolidated most of the TLS related utility code into a new file in the `util` package: `solr_tls_util.go`

All the TLS related config (env vars, annotations, initContainers, volumes/mounts) get appended to the config for a `StatefulSet` or `Deployment` in one step at the end vs. being infused at various places in `GenerateStatefulSet` and `GenerateSolrPrometheusExporterDeployment`. I like this approach because most of the TLS config is additive. There are a few places where we have to be careful about overriding previous settings (such as the `entrypoint` for the exporter when using the `mountedServerTLSDir` but that's only in a few small places).

There's also a lot of common logic for configuring the StatefulSet (for SolrClouds) and Deployment (for exporter), so I refactored the reusable parts into their own method.

### CODE REVIEW

File: controllers/controller_utils_test.go [hunk 1]
```diff
@@ -190,14 +190,6 @@ func verifyUserSuppliedTLSConfig(t *testing.T, tls *solr.SolrTLSOptions, expecte
...
 	assert.Equal(t, expectedTlsSecretName, tls.PKCS12Secret.Name)
 	assert.Equal(t, "keystore.p12", tls.PKCS12Secret.Key)
-
-	// is there a separate truststore?
-	expectedTrustStorePath := ""
-	if tls.TrustStoreSecret != nil {
-		expectedTrustStorePath = util.DefaultTrustStorePath + "/" + tls.TrustStoreSecret.Key
-	}
-
-	expectTLSEnvVars(t, util.TLSEnvVars(tls, needsPkcs12InitContainer), expectedKeystorePasswordSecretName, expectedKeystorePasswordSecretKey, needsPkcs12InitContainer, expectedTrustStorePath)
```
Comment (thelabdude) [inline controllers/controller_utils_test.go:200]: We don't need to call `expectTLSEnvVars` in this part of the test code as it already gets called after reconciliation, see `expectTLSConfigOnPodTemplate` and `expectMountedTLSDirConfigOnPodTemplate`. Removing this code allows us to hide the `TLSEnvVars` in the util package, it doesn't need to be exposed here.

File: controllers/solrcloud_controller.go [hunk 2]
```diff
@@ -215,15 +215,15 @@ func (r *SolrCloudReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 			// if there's a user-provided config, it must have one of the expected keys
 			if !hasLogXml && !hasSolrXml {
 				// TODO: Create event for the CRD.
-				return requeueOrNot, fmt.Errorf("User provided ConfigMap %s must have one of 'solr.xml' and/or 'log4j2.xml'",
+				return requeueOrNot, fmt.Errorf("user provided ConfigMap %s must have one of 'solr.xml' and/or 'log4j2.xml'",
```
Comment (thelabdude) [inline controllers/solrcloud_controller.go:218]: Just cleaning up some IDE nits here ... doesn't need to be in this PR but shouldn't hurt either ;-)

File: controllers/solrcloud_controller_tls_test.go [hunk 3]
```diff
@@ -126,7 +126,6 @@ func TestMountedTLSDir(t *testing.T) {
 	mountedDir := &solr.MountedTLSDirectory{}
 	mountedDir.Path = "/mounted-tls-dir"
 	instance.Spec.SolrTLS = &solr.SolrTLSOptions{MountedServerTLSDir: mountedDir, CheckPeerName: true, ClientAuth: "Need", VerifyClientHostname: true}
-	expectMountedTLSDirEnvVars(t, util.TLSEnvVars(instance.Spec.SolrTLS, false))
```
Comment (thelabdude) [inline controllers/solrcloud_controller_tls_test.go:129]: the TLS env vars will get checked in the `expectMountedTLSDirConfigOnPodTemplate` after reconciliation, so no need to do here ... same comment for change below this too ...

File: controllers/util/solr_util.go [hunk 4]
```diff
@@ -601,6 +552,11 @@ func GenerateStatefulSet(solrCloud *solr.SolrCloud, solrCloudStatus *solr.SolrCl
 		}
 	}
 
+	// Enrich the StatefulSet config to enable TLS on Solr pods if needed
+	if tls != nil {
+		tls.enableTLSOnSolrCloudStatefulSet(stateful)
```
Comment (thelabdude) [inline controllers/util/solr_util.go:557]: This is the crux of the design I took for this refactoring task. Basically, take a configured StatefulSet and then enrich it with all the TLS things in one go vs. spread throughout the method as it was before.

File: controllers/solrcloud_controller.go [hunk 5]
```diff
@@ -364,51 +364,57 @@ func (r *SolrCloudReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
-			if instance.Spec.SolrTLS.RestartOnTLSSecretUpdate {
-				if tlsCertBytes, ok := foundTLSSecret.Data[util.TLSCertKey]; ok {
-					tlsCertMd5 = fmt.Sprintf("%x", md5.Sum(tlsCertBytes))
-				} else {
-					return requeueOrNot, fmt.Errorf("%s key not found in TLS secret %s, cannot watch for updates to"+
-						" the cert without this data but 'solrTLS.restartOnTLSSecretUpdate' is enabled!",
-						util.TLSCertKey, foundTLSSecret.Name)
+	var tls *util.TLSConfig
+	if !blockReconciliationOfStatefulSet && instance.Spec.SolrTLS != nil {
+		tls = &util.TLSConfig{}
```
Comment (thelabdude) [inline controllers/solrcloud_controller.go:370]: The `TLSConfig` struct allows us to hold the `TLSOptions` that from from the user config as well as additional config info determined during reconciliation, such as the MD5 hash of the cert. Not married to the name of this struct ... could be `TLSOptionsAndReconciledVars`

File: controllers/solrprometheusexporter_controller.go [hunk 6]
```diff
@@ -197,28 +197,30 @@ func (r *SolrPrometheusExporterReconciler) Reconcile(req ctrl.Request) (ctrl.Res
 					prometheusExporter.Spec.SolrReference.SolrTLS.KeyStorePasswordSecret.Key, keyStorePasswordSecret.Name)
 			}
 
-			tlsClientOptions = &util.TLSClientOptions{}
```
Comment (thelabdude) [inline controllers/solrprometheusexporter_controller.go:200]: The `TLSClientOptions` name was misleading once I repurposed this struct to be used by the exporter and the StatefulSet code

File: controllers/util/prometheus_exporter_util.go [hunk 7]
```diff
@@ -46,17 +45,9 @@ type SolrConnectionInfo struct {
 	StandaloneAddress      string
 }
 
-// Used internally to capture config needed to provided Solr client apps like the exporter
-// with config needed to call TLS enabled Solr pods
-type TLSClientOptions struct {
```
Comment (thelabdude) [inline controllers/util/prometheus_exporter_util.go:51]: Replaced by `TLSConfig` in `solr_tls_util.go`

File: controllers/util/solr_util.go (hunk 4 above)
Comment (HoustonPutman) [reply controllers/util/solr_util.go:557]: Yeah this is great. Hopefully we can abstract out other parts of this awful method and make it nice and clean.

File: controllers/solrcloud_controller.go (hunk 5 above)
Comment (HoustonPutman) [reply controllers/solrcloud_controller.go:370]: `TLSConfig` is fine with me

**Review by HoustonPutman — APPROVED** [review]
I really like the new design, this looks good to me!
